### ISSUE TITLE
Streams don't quite make up for v2's Query.fold

### ISSUE THREAD

**Issue by qualidafial** [jdbi/jdbi#526]
In the process of converting my v2 project to v3, I've repeatedly found myself wishing to have `Query.fold` back--or something similar to it.

When we removed it, the idea was that Java 8 streams would fill the niche. They're very handy, but not all use cases are served by it. In particular:

* For model-based folds (`JoinRowMapper` is great for this), `Stream.reduce(U, BiFunction<U,T,U>, BinaryOperator<U>)` does the job, but the `BinaryOperator` parameter is pointless, since this is not a parallel stream. This is one instance where I think the JDK over-engineered streams for parallelism, which hurts usability for sequential streams.
* We still need a fold with raw access to the raw `ResultSet`. Not all use cases fit in the map-reduce approach.

I propose we add the following API:

```java
@FunctionalInterface
interface ResultSetAccumulator<U> {
    U reduce(U identity, ResultSet rs, StatementContext ctx);
}

class Query {
    ...
    <U> U reduce(U identity, ResultSetAccumulator<U> accumulator);
    <U> U reduce(U identity, BiFunction<U, T, U> accumulator);
}
```

**Comment by stevenschlansker** [comment]
I'll give this one a go, unless you've already started.

**Comment by qualidafial** [comment]
All yours

**Comment by qualidafial** [comment]
I guess `ResultBearing` is a better home for the new methods.

**Comment by umcodemonkey** [comment]
Random question from the peanut gallery: does `Stream.map(Function<T, U>).reduce(U, BinaryOperator<U>)` not achieve the same goal? Or are there use cases where this would produce incorrect results? Not considering the awkward syntax here, just correctness.

**Comment by umcodemonkey** [comment]
Never mind; I missed the comment about needing access to the raw `ResultSet`. Probably could be worked around, but inconvenient to do so.

**Comment by qualidafial** [comment]
@umcodemonkey it depends on the use case. If your row type is the same as a reduced type then `Stream.reduce(U, BinaryOperator<U>)` works fine.

However in my case my result type is generally some kind of collection. e.g. say your row type is `JoinRow` containing a `Foo` and `Bar`, and you want a `ListMultimap<Foo,Bar>` back from the method. That's when the `BinaryOperator` parameter (in this case, a `BinaryOperator<ListMultimap<Foo,Bar>>`) is superfluous. It's only for parallelism, and our stream is strictly sequential.

**Comment by umcodemonkey** [comment]
It may be my ignorance, but that *specific* use case looks like a `map`/`collect` to my eyes, with the problem that there probably isn't an out-of-the box for `ListMultimap` collector to make that not a terrible exercise in "why is this so much harder than it has to be!" :+1: for providing better interfaces.

**Comment by stevenschlansker** [comment]
I think the `ResultSetAccumulator` interface actually turns out to not work so well.  By the time you have a `ResultBearing<T>`, you have already configured mapping the type `<T>`.  `ResultBearing` only has `<T> iterator()` which hides the `ResultSet`.

Should we consider adding `ResultSet` as a "mappable" type, which just repeatedly emits the single `ResultSet` cursor and trusts the user to do the right thing?  Then you can do whatever you want with it via any of the supported interfaces.  Crazy idea: we could even make it the default mapped type.  The `Map<String, Object>` default sounds nice in theory but it's weird since databases fold case a lot and allocating all those strings / maps is pretty costly.

**Comment by stevenschlansker** [comment]
I sent a PR with the first part for discussion.  I'm holding off on the `ResultSetAccumulator` half until we discuss the point above.

**Comment by stevenschlansker** [comment]
Hm, the idea of `.mapTo(ResultSet.class)` is weakened a lot by the checked `SqlException` that keeps you from making lambdas :cry:

**Comment by qualidafial** [comment]
I'm ok with pulling `fold(U, ResultSetAccumulator<U>)` up to `Query` if that's what's holding it back.

**Comment by stevenschlansker** [comment]
That introduces a new query execution strategy, and would require either changing `execute()` or introducing a similar-but-different version.  I'm not necessarily against it, but could you maybe enumerate a couple of use cases which don't fit map-reduce and need this escape hatch?

**Comment by qualidafial** [comment]
Honestly, a lot of it is sheer irritation at having to invent one-off classes just to hold a single field:

```java
public class FinishProfileStep {
  final Long stepId;
  public FinishProfileStep(Long stepId) {
    this.stepId = stepId;
  }
}

List<FinishProfile> list() {
  Map<Long, FinishStep> steps = uniqueIndex(
      getHandle().attach(FinishStepDao.class).list(),
      FinishStep::getId);

  return getHandle().createQuery("select profile.id, profile.name, step.step_id "
          + "from finish_profile profile "
          + "left join finish_profile_step step on profile.id = step.profile_id "
          + "order by profile.name")
      .registerMapper(ConstructorMapper.of(FinishProfileStep.class))
      .map(JoinRowMapper.forTypes(FinishProfile.class, FinishProfileStep.class))
      .reduce(new LinkedHashMap<Long, FinishProfile>(), (map, joinRow) -> {
        FinishProfile profile = map.computeIfAbsent(
            joinRow.get(FinishProfile.class).getId(),
            id -> joinRow.get(FinishProfile.class));
        Optional.of(joinRow.get(FinishProfileStep.class).stepId)
            .map(steps::get)
            .ifPresent(step -> profile.getSteps().add(step);
        return map;
      })
      .values()
      .stream()
      .collect(toList());
}
```

vs

```java
List<FinishProfile> list() {
  Map<Long, FinishStep> steps = uniqueIndex(
      getHandle().attach(FinishStepDao.class).list(),
      FinishStep::getId);

  return getHandle().createQuery("select profile.id, profile.name, step.step_id "
          + "from finish_profile profile "
          + "left join finish_profile_step step on profile.id = step.profile_id "
          + "order by profile.name")
      .reduce(new LinkedHashMap<Long, FinishProfile>(), (rs, ctx) -> {
        long profileId = rs.getLong("id");
        String name = rs.getString("name");
        FinishProfile profile = map.computeIfAbsent(profileId, id -> new FinishProfile(id, name))
        Optional.of(rs.getLong("step_id"))
            .map(steps::get)
            .ifPresent(step -> profile.getSteps().add(step);
      })
      .values()
      .stream()
      .collect(toList());
}
```

**Comment by stevenschlansker** [comment]
Okay, that is a bit of a bummer.  Do you agree with the above musings, namely that it's actually relatively difficult to get this in without adding new execution strategies?

We may have to improve the `execute()` method to take some hooks so the map step can be customized.

**Comment by qualidafial** [comment]
It may well require a new execution strategy; I haven't looked closely at implementation yet. For now, I'm focusing on how the interface feels from the user's perspective.

**Comment by qualidafial** [comment]
Hmm.. we could probably cheat and just map to a ResultSet/context pair, then wrap the ResultSetAccumulator in a BiFunction that catches and rethrows SQLExceptions. Oh my god that looks like complete nonsense without context.

**Comment by qualidafial** [comment]
Fixed in #540
